cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaListenersReconciler.java

```diff
@@ -442,6 +443,63 @@ protected Future<Void> internalServicesReady()   {
         return Future.succeededFuture();
     }
 
+    /**
+     * Collect all addresses of services related to clusterIP addresses for Statuses, certificates and advertised
+     * addresses. This method for all ClusterIP type listeners:
+     *      1) Collects the relevant addresses of bootstrap service and stores them for use in certificates
+     *      2) Collects the clusterIP addresses for certificates and advertised hostnames
+     *
+     * @return  Future which completes clusterIP service addresses are collected
+     */
+    protected Future<Void> clusterIPServicesReady()   {
+        for (GenericKafkaListener listener : ListenersUtils.clusterIPListeners(kafka.getListeners())) {
+            boolean useServiceDnsDomain = (listener.getConfiguration() != null && listener.getConfiguration().getUseServiceDnsDomain() != null)
+                    ? listener.getConfiguration().getUseServiceDnsDomain() : false;
+
+            String bootstrapAddress = listener.getConfiguration().getBootstrap().getHost();
+
+            if (bootstrapAddress == null) {
+                // Set status based on bootstrap service
+                bootstrapAddress = getInternalServiceHostname(reconciliation.namespace(), ListenersUtils.backwardsCompatibleBootstrapServiceName(reconciliation.name(), listener), useServiceDnsDomain);
+            }
+
+            if (listener.isTls()) {
+                result.bootstrapDnsNames.add(bootstrapAddress);
+            }
+
+            ListenerStatus ls = new ListenerStatusBuilder()
+                    .withName(listener.getName())
+                    .withAddresses(new ListenerAddressBuilder()
+                            .withHost(bootstrapAddress)
+                            .withPort(listener.getPort())
+                            .build())
+                    .build();
+            result.listenerStatuses.add(ls);
+
+            // Set advertised hostnames and ports
+            for (int brokerId = 0; brokerId < kafka.getReplicas(); brokerId++) {
+                String brokerAddress;
+
+                if (useServiceDnsDomain) {
+                    brokerAddress = DnsNameGenerator.podDnsNameWithClusterDomain(reconciliation.namespace(), KafkaResources.brokersServiceName(reconciliation.name()), KafkaResources.kafkaStatefulSetName(reconciliation.name()) + "-" + brokerId);
+                } else {
+                    brokerAddress = DnsNameGenerator.podDnsNameWithoutClusterDomain(reconciliation.namespace(), KafkaResources.brokersServiceName(reconciliation.name()), KafkaResources.kafkaStatefulSetName(reconciliation.name()) + "-" + brokerId);
```

Comment:
This looks like it would not direct the traffic through the ClusterIP service but instead try to use it as headless service. I think you need to use here `getInternalServiceHostname` method as well. Just feed it the name of the per-broker service, which I guess is generated through `ListenersUtils.backwardsCompatibleBrokerServiceName`.